changefeedccl/resolvedspan: integrate span.MultiFrontier into frontiers

andyyang890 opened the pull request:
Fixes #148110

---

**changefeedccl: fix ALTER CHANGEFEED span selection** 

This patch updates the ALTER CHANGEFEED code to use primary index spans
instead of entire table spans to be consistent with the rest of the
changefeed code and to help maintain the assumption that we will
never merge spans from different tables in our frontier data structures.

Release note: None

---

**changefeedccl/resolvedspan: integrate span.MultiFrontier into frontiers** 

This patch integrates `span.MultiFrontier`s into the resolved span frontiers
to give changefeed processors easy access to per-table resolved timestamps.

Release note: None

---

**changefeedccl/resolvedspan: add feature flag for per-table tracking** 

This patch adds a new cluster setting named
`changefeed.frontier.per_table_tracking.enabled`
to feature flag enabling per-table tracking in
the resolved span frontiers.

Release note: None

andyyang890 committed:
changefeedccl/resolvedspan: integrate span.MultiFrontier into frontiers

This patch integrates `span.MultiFrontier`s into the resolved span frontiers
to give changefeed processors easy access to per-table resolved timestamps.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151355)
<!-- Reviewable:end -->

andyyang890 committed:
changefeedccl: fix ALTER CHANGEFEED span selection

This patch updates the ALTER CHANGEFEED code to use primary index spans
instead of entire table spans to be consistent with the rest of the
changefeed code and to help maintain the assumption that we will
never merge spans from different tables in our frontier data structures.

Release note: None

aerfrei reviewed (approved):
Looks good! Nice and clean, thanks for this.

Another couple of more general things I want to confirm:
- After this change, all changefeeds, including ones watching only a single table will use this multiFrontier, but that multiFrontier will only ever have one subFrontier
- The frontier objects only exist in memory and are recreated from the stored job progress on changefeed restart. No additional work will be needed to convert between the frontier and multiFrontier e.g. on upgrades.

asg0451 commented on pkg/ccl/changefeedccl/resolvedspan/frontier.go:
```diff
@@ -184,15 +191,11 @@ func (f *CoordinatorFrontier) MakeCheckpoint(
 // lagging and checkpoint spans.
 type resolvedSpanFrontier struct {
-	spanFrontier
+	*span.MultiFrontier[descpb.ID]
```
do you think it's worth feature-flagging this?

asg0451 commented on pkg/ccl/changefeedccl/alter_changefeed_stmt.go:
```diff
@@ -541,12 +541,16 @@ func generateAndValidateNewTargets(
+			var existingTargetSpanIDs []spanID
 			for _, targetDesc := range newTableDescs {
-				existingTargetIDs = append(existingTargetIDs, targetDesc.GetID())
+				tableDesc := targetDesc.(catalog.TableDescriptor)
```
nit: assertionfailed rather than panic (unless you're really really sure)

andyyang890 committed:
changefeedccl/resolvedspan: add feature flag for per-table tracking

This patch adds a new cluster setting named
`changefeed.frontier.per_table_tracking.enabled`
to feature flag enabling per-table tracking in
the resolved span frontiers.

Release note: None

andyyang890 commented:
TFTRs!

bors r=aerfrei,asg0451
